File: internal/api/app_status_test.go
package api

import (
	"testing"

	"github.com/GLINCKER/levelrail/internal/reconcile"
)

func TestSummarizeAppConditions(t *testing.T) {
	tests := []struct {
		name       string
		conditions []reconcile.Condition
		want       appStatusSummary
	}{
		{
			name:       "no conditions",
			conditions: nil,
			want:       appStatusSummary{Label: "No status yet", Variant: "muted"},
		},
		{
			name: "all true is healthy",
			conditions: []reconcile.Condition{
				{Type: "Ready", Status: reconcile.ConditionTrue, Reason: "Deployed"},
			},
			want: appStatusSummary{Label: "Healthy", Variant: "success"},
		},
		{
			name: "a false condition needs attention regardless of others",
			conditions: []reconcile.Condition{
				{Type: "Ready", Status: reconcile.ConditionTrue, Reason: "Deployed"},
				{Type: "EgressPolicyReady", Status: reconcile.ConditionFalse, Reason: "PolicyApplyFailed"},
			},
			want: appStatusSummary{Label: "Attention needed", Variant: "destructive"},
		},
		{
			name: "suspended overrides everything else",
			conditions: []reconcile.Condition{
				{Type: "Ready", Status: reconcile.ConditionUnknown, Reason: "Suspended"},
			},
			want: appStatusSummary{Label: "Stopped", Variant: "muted"},
		},
		{
			name: "an unconfigured optional feature does not block healthy",
			conditions: []reconcile.Condition{
				{Type: "Ready", Status: reconcile.ConditionTrue, Reason: "AlreadyRunning"},
				{Type: "EgressPolicyReady", Status: reconcile.ConditionUnknown, Reason: "NotConfigured"},
			},
			want: appStatusSummary{Label: "Healthy", Variant: "success"},
		},
		{
			name: "a disabled optional integration does not block healthy",
			conditions: []reconcile.Condition{
				{Type: "Ready", Status: reconcile.ConditionTrue, Reason: "Deployed"},
				{Type: "Ready", Status: reconcile.ConditionUnknown, Reason: "Disabled"},
			},
			want: appStatusSummary{Label: "Healthy", Variant: "success"},
		},
		{
			name: "a genuinely pending unknown condition still reconciles",
			conditions: []reconcile.Condition{
				{Type: "Ready", Status: reconcile.ConditionUnknown, Reason: "AwaitingFirstBuild"},
			},
			want: appStatusSummary{Label: "Reconciling", Variant: "muted"},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := summarizeAppConditions(tt.conditions); got != tt.want {
				t.Errorf("summarizeAppConditions() = %+v, want %+v", got, tt.want)
			}
		})
	}
}

File: web/src/lib/appStatus.test.ts
import { describe, expect, it } from 'vitest'
import { summarizeAppStatus } from './appStatus'
import type { ReconcileCondition } from '../types/deploy'

function condition(overrides: Partial<ReconcileCondition>): ReconcileCondition {
  return {
    Type: 'Ready',
    Status: 'True',
    Reason: 'AlreadyRunning',
    Message: '',
    LastTransitionTime: '2026-01-01T00:00:00Z',
    ...overrides,
  }
}

describe('summarizeAppStatus', () => {
  it('reports no status yet for an empty conditions list', () => {
    expect(summarizeAppStatus([])).toEqual({
      label: 'No status yet',
      variant: 'muted',
    })
  })

  it('reports healthy when every condition is true', () => {
    expect(summarizeAppStatus([condition({})])).toEqual({
      label: 'Healthy',
      variant: 'success',
    })
  })

  it('reports attention needed for any false condition regardless of others', () => {
    const result = summarizeAppStatus([
      condition({}),
      condition({ Type: 'EgressPolicyReady', Status: 'False' }),
    ])
    expect(result).toEqual({
      label: 'Attention needed',
      variant: 'destructive',
    })
  })

  // The exact regression found via real deployment testing: an app with
  // no egress policy configured (the common case for virtually every
  // app) carries a permanently-Unknown EgressPolicyReady condition that
  // must not keep an otherwise-healthy app stuck on "Reconciling".
  it('reports healthy when the only non-true condition is an unconfigured optional feature', () => {
    const result = summarizeAppStatus([
      condition({}),
      condition({
        Type: 'EgressPolicyReady',
        Status: 'Unknown',
        Reason: 'NotConfigured',
      }),
    ])
    expect(result).toEqual({ label: 'Healthy', variant: 'success' })
  })

  it('reports healthy when the only non-true condition is a disabled optional integration', () => {
    const result = summarizeAppStatus([
      condition({}),
      condition({ Type: 'Ready', Status: 'Unknown', Reason: 'Disabled' }),
    ])
    expect(result).toEqual({ label: 'Healthy', variant: 'success' })
  })

  it('still reports reconciling for a genuinely pending unknown condition', () => {
    const result = summarizeAppStatus([
      condition({ Status: 'Unknown', Reason: 'AwaitingFirstBuild' }),
    ])
    expect(result).toEqual({ label: 'Reconciling', variant: 'muted' })
  })

  it('reports stopped when any condition reports Suspended, overriding everything else', () => {
    const result = summarizeAppStatus([
      condition({ Reason: 'Suspended' }),
      condition({ Type: 'EgressPolicyReady', Status: 'False' }),
    ])
    expect(result).toEqual({ label: 'Stopped', variant: 'muted' })
  })
})
